PWN-V1.3 | Ethereum | 24.02.2025

Vojtch159 opened the pull request:
gm, submitting PWN for a protocol review. Let me know if everything's okay and if there's anything I can do to help with the review. Thanks!

https://pwn.xyz/
https://app.pwn.xyz/
https://docs.pwn.xyz/
https://dev-docs.pwn.xyz/
https://github.com/PWNDAO/pwn_contracts

emduc commented on src/content/protocols/pwn.md:
```diff
@@ -0,0 +1,140 @@
+`TransparentProxy` contract with the PWN DAO as admin is used for the `PWNConfig`. However, this does not impact user funds or otherwise materially change the protocol for already existing loans. Only new loans are affected by the change of parameters in the `PWNConfig`.
+
+There's also a timelock for the upgrades made by the DAO which ultimatelly owns `PWNConfig` and `PWNHub` that is currently set to 0 days.
+
+> Upgradeabillity score: M
+
```
If on user funds nor unclaimed yield are at risk, this qualifies as low risk.

emduc commented on src/content/protocols/pwn.md:
```diff
@@ -0,0 +1,140 @@
+> Exit score: L
+
+## Accessibility
+
+PWN can be used on the PWN Platform (app.pwn.xyz). Once a loan is created, interactions with the protocol are simple enough to be feasibly done through etherscan in case of the main interface failure.
+
+> Accessibility score: M
```
Are there instructions in the docs to recover the funds using etherscan? If that's the case it would be great to include a link. If not, unfortunately it would be a High risk score.

Vojtch159 replied:
Added, thanks.

emduc commented on src/content/protocols/pwn.md:
```diff
@@ -0,0 +1,140 @@
+| PWNUtilizedCredit | utilizeCredit | Updates utilised credit for an account. This is used to share credit between different proposal types. Malicious actor could update the utilised credit to go over the credit limit. | Contract with a valid tag in PWNHub |
+
+
+## Dependencies
+
+No external dependency has been found.
+
```
The docs mention Chainlink Oracles for the elastic chainlink proposals. Could you quickly explain here how chainlink is used.

emduc commented on src/content/protocols/pwn.md:
```diff
@@ -0,0 +1,140 @@
+
+# Technical Analysis
+
+## Contracts
+
+| Contract Name                               | Address                                    |
```
could you please include the governance contracts to your analysis and the timelock contract that holds the permissions.

Vojtch159 replied:
https://github.com/deficollective/defiscan/pull/102#issuecomment-3084296151

emduc commented on src/content/protocols/pwn.md:
```diff
@@ -0,0 +1,140 @@
+| ------------- | ------------ | ----------- | ----------------------- |
+| TransparentUpgradeableProxy | fallback | Fallback is called when no other function matches the call data. No specific malicious action can be done if admin is compromised. | TimelockController |
+| TransparentUpgradeableProxy | receive | Same as fallback but receive is called when calldata are empty. No specific malicious action can be done if admin is compromised. | TimelockController |
+| TransparentUpgradeableProxy | _fallback | Same as fallback but when caller is the admin. No specific malicious action can be done if admin is compromised. | TimelockController |
```
I believe this function is also used for the implementation upgrades, right?

Vojtch159 replied:
oh thanks for the catch, fixed

emduc commented on src/content/protocols/pwn.md:
```diff
@@ -0,0 +1,140 @@
+| PWNHub | setTags | Same as setTag. This function calls setTag multiple times. | TimelockController |
+| PWNLOAN | mint | Mints a LOAN token. LOAN token represents a loan in the protocol for the lender. It is required to claim repayment or collateral. Malicious actor could mint worthless LOAN tokens and try to use them as a collateral for loans. | Loan contract specified in an offer |
+| PWNLOAN | burn | Burns a LOAN token.LOAN token represents a loan in the protocol for the lender. It is required to claim repayment or collateral. Malicious actor could burn other LOAN tokens and make it impossible to claim loans. | Loan contract specified in an offer |
+| PWNRevokedNonce | revokeNonce | Revokes nonce for a offer. Used to revoke an offchain signed offer on behalf of an owner. Malicious actor could revoke offers that the user wants to be valid. | Contract with a valid tag in PWNHub |
```
typo: for an offer

emduc commented on src/content/protocols/pwn.md:
```diff
@@ -0,0 +1,140 @@
+| PWNLOAN | mint | Mints a LOAN token. LOAN token represents a loan in the protocol for the lender. It is required to claim repayment or collateral. Malicious actor could mint worthless LOAN tokens and try to use them as a collateral for loans. | Loan contract specified in an offer |
+| PWNLOAN | burn | Burns a LOAN token.LOAN token represents a loan in the protocol for the lender. It is required to claim repayment or collateral. Malicious actor could burn other LOAN tokens and make it impossible to claim loans. | Loan contract specified in an offer |
+| PWNRevokedNonce | revokeNonce | Revokes nonce for a offer. Used to revoke an offchain signed offer on behalf of an owner. Malicious actor could revoke offers that the user wants to be valid. | Contract with a valid tag in PWNHub |
+| PWNRevokedNonce | revokeNonce | Revokes nonce for a offer. Used to revoke an offchain signed offer on behalf of an owner. Malicious actor could revoke offers that the user wants to be valid. | Contract with a valid tag in PWNHub |
```
typo: for an offer

emduc reviewed (changes requested):
Hi there!

Thanks a lot for your submission and sorry for the delay.

I left a few comments here and there but overall the quality of the review is quite good. We appreciate the effort you made in the analysis of permissions to highlight the potential attack vectors and consequences.

Could you look into the compilation issues so that it can compile and render in pre-review?

Here are a few additions that we would find useful:

- Could you add the governance contracts to the list?
- We recently started publishing diagrams with our reviews to help users understand the system better. You can see a good example of diagram with our recent Compound V3 review [here ](https://www.defiscan.info/protocols/compound-v3). Would you be open to designing a diagram too? We created a canvas with the basics [here](https://link.excalidraw.com/l/9pt8PDVB43r/3rqeXJqhsGF).
- We also decided to add a small conclusion section at the end of the overview, to summarize the findings in 1-2 sentences and explain the overall stage score.

Thanks again for your submission and we're looking forward to those small improvements!

emduc commented:
Hey there! @Vojtch159 

Did you have a chance to look at our comments? We'd be excited to publish the review

Vojtch159 commented:
gm, I've updated the review based on the feedback and added the new protocol contract. 

I added an explanation of our governance setup in the Governance section but I haven't added the governance contracts as we're using the Aragon OSx framework. I think it would make more sense to add a review of the framework separately in the future, similarly to Chainlink Price feeds review - is that okay? Let me know if some information about the governance parameters is unclear.

Reg. protocol diagram - there already exists one in our docs so I used it to avoid duplicates.

emduc reviewed (commented):
Hi!

Thanks a lot for the updates, it's a great submission. Here are a few points to address.

 - I would just like to clarify the permission owners. Currently there seems to be no Stewards that have a relevant role in the protocol?
 - Your diagram shows an additional Admin timelock, but there is only one timelcok in the permission owners table.
 - I scanned the Timelock that is currently listed and found that it has no mandatory delay, and the following roles:

  "access_control_roles": {
    "TIMELOCK_ADMIN_ROLE": ["0xd8dbddf1c0fddf9b5ecfa5c067c38db66739fbab"],
    "PROPOSER_ROLE": ["0x1b8383d2726e7e18189205337424a2631a2102f4"],
    "CANCELLER_ROLE": ["0x1b8383d2726e7e18189205337424a2631a2102f4"],
    "EXECUTOR_ROLE": ["0x0000000000000000000000000000000000000000"]
  }

That would yield the following permission to add to the table:

| TimelockController | schedule|  Schedules a transaction to be executed after the mandatory delay has passed (currently none). This transaction can perform any action that this contract is allowed to, including upgrading `PwnConfig` and tagging new contracts in `PwnHub`. | PwnDAO |
| TimelockController | scheduleBatch| Similar to`schedule`, for a batch of transactions. | PwnDAO |
| TimelockController | updateDelay |  Updates the mandatory delay. There are no minimum or maximum delays enforced in the contract. | TimelockController  |
| TimelockController | cancel |  Cancels a queued transaction before it has been executed. | PwnDAO |
| TimelockController | grantRole | Grants a role to a given address. There are different roles to schedule, cancel, and execute transactions, currently executing transactions is open to anyone. | TimelockController  |
| TimelockController | revokeRole| Revokes a role from a given address.  | TimelockController  |


 - The multi token registry that is on the diagram is also not mentioned anywhere in this review, is it a recent addition?
 - Finally, I adapted the diagram to meet out usual designs on excalidraw:

<img width="2507" height="1716" alt="Image" src="https://github.com/user-attachments/assets/70ca6ef6-dfb4-485c-a0a4-665c431a7a25" />

You can edit it freely here: https://link.excalidraw.com/l/9pt8PDVB43r/3rqeXJqhsGF

yvesbou commented on src/content/protocols/pwn/ethereum.md:
```diff
@@ -0,0 +1,197 @@
+| PWNLOAN | burn | Burns a LOAN token.LOAN token represents a loan in the protocol for the lender. It is required to claim repayment or collateral. Malicious actor could burn other LOAN tokens and make it impossible to claim loans. | Loan contract specified in an offer |
+| PWNRevokedNonce | revokeNonce | Revokes nonce for an offer. Used to revoke an offchain signed offer on behalf of an owner. Malicious actor could revoke offers that the user wants to be valid. | Contract with a valid tag in PWNHub |
+| PWNRevokedNonce | revokeNonce | Revokes nonce for an offer. Used to revoke an offchain signed offer on behalf of an owner. Malicious actor could revoke offers that the user wants to be valid. | Contract with a valid tag in PWNHub |
+| PWNSimpleLoan | createLOAN | Creates a loan in the PWN Protocol. The check in this function is for a valid lenderSpec check in case the caller is a lender. There's also a reentrancy protection modifier. | Lender |
+| PWNSimpleLoan | repayLOAN | Repays a running loan. There's a reentrancy protection modifier to make sure you can't repay more than once. | Borrower |
+| PWNSimpleLoan | claimLOAN | Claims repayment or defaulted collateral. There's a reentrancy protection modifier to make sure you can't claim more than once. | Lender |
+| PWNSimpleLoan | tryClaimRepaidLOAN | Tries to claim loan for a lender. The check is this function is to make sure this function can only be called by the Loan contract. Malicious actor could initiate the claim without the Loan contract. Note that it's not possible to claim through this function if the original loan lender has changed and therefor malicious actor cannot claim the assets to its w ...
+| PWNSimpleLoan | makeExtensionProposal | Makes an onchain proposal for loan extension. The borrower can create a proposal for extension. Malicious actor could create an extension proposal for any loan. | Proposer |
+| PWNSimpleLoan | extendLOAN | Extends a loan based on extension proposal. Both the lender and the borrower can call this function to extend a loan they're part of if the other party has signed the extension proposal. The checks in this function are to distinguish two possible states (lender or borrower calling the function). | Lender or borrower |
```
we dont include user actions in protocols as part of the permissions, we didnt include user actions in aave, spark and pendle e.g
```suggestion
```

yvesbou commented on src/content/protocols/pwn/ethereum.md:
```diff
@@ -0,0 +1,197 @@
+| PWNSimpleLoan | tryClaimRepaidLOAN | Tries to claim loan for a lender. The check is this function is to make sure this function can only be called by the Loan contract. Malicious actor could initiate the claim without the Loan contract. Note that it's not possible to claim through this function if the original loan lender has changed and therefor malicious actor cannot claim the assets to its w ...
+| PWNSimpleLoan | makeExtensionProposal | Makes an onchain proposal for loan extension. The borrower can create a proposal for extension. Malicious actor could create an extension proposal for any loan. | Proposer |
+| PWNSimpleLoan | extendLOAN | Extends a loan based on extension proposal. Both the lender and the borrower can call this function to extend a loan they're part of if the other party has signed the extension proposal. The checks in this function are to distinguish two possible states (lender or borrower calling the function). | Lender or borrower |
+| PWNSimpleLoanSimpleProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
+| PWNSimpleLoanSimpleProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
+| PWNSimpleLoanSimpleProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
+| PWNSimpleLoanSimpleProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
+| PWNSimpleLoanListProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
+| PWNSimpleLoanListProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
+| PWNSimpleLoanListProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
+| PWNSimpleLoanListProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
+| PWNSimpleLoanElasticProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
+| PWNSimpleLoanElasticProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
+| PWNSimpleLoanElasticProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
+| PWNSimpleLoanElasticProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
+| PWNSimpleLoanDutchAuctionProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
+| PWNSimpleLoanDutchAuctionProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
+| PWNSimpleLoanDutchAuctionProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
+| PWNSimpleLoanDutchAuctionProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
+| PWNSimpleLoanElasticChainlinkProposal | _makeProposal | Marks a proposalHash as valid. Caller has to also be a proposer of the corresponding offer. | Proposer |
+| PWNSimpleLoanElasticChainlinkProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
+| PWNSimpleLoanElasticChainlinkProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
+| PWNSimpleLoanElasticChainlinkProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
```
```suggestion
| PWNSimpleLoanSimpleProposal | acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
| PWNSimpleLoanListProposal | acceptProposal |Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
| PWNSimpleLoanElasticProposal | acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
| PWNSimpleLoanDutchAuctionProposal | acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
| PWNSimpleLoanElasticChainlinkProposal | acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. | Loan contract specified in an offer |
```

yvesbou replied:
summarise _acceptProposal and acceptProposal, remove user only facing external/public endpoints

yvesbou commented on src/content/protocols/pwn/ethereum.md:
```diff
@@ -0,0 +1,197 @@
+| PWNSimpleLoanElasticChainlinkProposal | _acceptProposal | Accepts a given proposal. Loan contracts call this function with the signed proposal on behalf of the acceptor. Malicious actor could accept proposals for different Loan contracts. | Loan contract specified in an offer |
+| PWNSimpleLoanElasticChainlinkProposal | makeProposal | Makes an onchain proposal. Checks in this function are derived from _makeProposal. | Proposer |
+| PWNSimpleLoanElasticChainlinkProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
+| PWNUtilizedCredit | utilizeCredit | Updates utilised credit for an account. This is used to share credit between different proposal types. Malicious actor could update the utilised credit to go over the credit limit. | Contract with a valid tag in PWNHub |
```
```suggestion
| PWNUtilizedCredit | utilizeCredit | Updates utilised credit for an account. This is used to share credit between different proposal types. Malicious actor would add a valid tag to a malicious contract which could update the utilised credit to go over the credit limit. | Contract with a valid tag in PWNHub |
```

yvesbou commented on src/content/protocols/pwn/ethereum.md:
```diff
@@ -0,0 +1,197 @@
+| PWNSimpleLoanElasticChainlinkProposal | acceptProposal | Accepts a given proposal. Checks in this function are derived from _acceptProposal. | Loan contract specified in an offer |
+| PWNUtilizedCredit | utilizeCredit | Updates utilised credit for an account. This is used to share credit between different proposal types. Malicious actor could update the utilised credit to go over the credit limit. | Contract with a valid tag in PWNHub |
+| TimelockController (Protocol) | schedule | Schedules a transaction to be executed after the mandatory delay has passed (currently none). This transaction can perform any action that this contract is allowed to, including upgrading PWNConfig and tagging new contracts in PWNHub. | PWNDAO |
+| TimelockController (Protocol) | scheduleBatch | Similar toschedule, for a batch of transactions. | PWNDAO |
```
```suggestion
| TimelockController (Protocol) | scheduleBatch | Similar to schedule, for a batch of transactions. | PWNDAO |
```

yvesbou commented on src/content/protocols/pwn/ethereum.md:
```diff
@@ -0,0 +1,197 @@
+| TimelockController (Protocol) | grantRole | Grants a role to a given address. There are different roles to schedule, cancel, and execute transactions, currently executing transactions is open to anyone. | TimelockController (Protocol) |
+| TimelockController (Protocol) | revokeRole| Revokes a role from a given address. | TimelockController (Protocol) |
+| TimelockController (Admin) | schedule | Schedules a transaction to be executed after the mandatory delay has passed (currently none). This transaction can perform any action that this contract is allowed to, including changing parameters in PWNConfig. | PWNDAO |
+| TimelockController (Admin) | scheduleBatch | Similar toschedule, for a batch of transactions. | PWNDAO |
```
```suggestion
| TimelockController (Admin) | scheduleBatch | Similar to schedule, for a batch of transactions. | PWNDAO |
```

Vojtch159 commented on src/content/protocols/pwn/ethereum.md:
```diff
@@ -0,0 +1,197 @@
+
+# Protocol Analysis
+
+![Overview of the PWN protocol](../diagrams/pwn-overview.png)
```
thanks, fixed.
